README.md
UniMorph canonicalization
=========================

**🚧 work in progress 👷**

This directory contains code for canonicalizing and validating
[UniMorph](https://unimorph.github.io/) feature bundles according to the
[official
documentation](www.unimorph.org/doc/Sylak-Glassman_2016_-_UniMorph_Schema_User_Guide.pdf).
The included command-line tool `um_canonicalize` converts UniMorph files
(three-column TSV files consisting of lemma, inflection, and feature bundle) so
that the features are placed in a canonical order (see below). This tool is
written in Pure Python and has no external dependencies except setuptools and
YAML.

Command-line tool
-----------------

### Installation

This package requires Python 3.6+.

You can install the package from GitHub directly using the following command:

``` {.bash}
pip install git+https://github.com/unimorph/um-canonicalize.git
```

### Usage

After installation, the terminal command `um_canonicalize` will be available.
For instance, if you have the modern Greek UniMorph file `ell` in your working
directory, you can issue:

``` {.bash}
um_canonicalize ell
```

This should enforce the canonicalization (described below) on the file, failing
if any conflicts or inconsistencies are detected; running this tool on its own
(successful) output should print "0 feature bundles canonicalized".

Canonical order
---------------

We propose a canonical ordering of tags within a feature bundle as follows.

-   The first tag is the "Part of Speech" tag.
-   The remaining universal tags are then placed in lexicographic order of the
    category they represent; e.g., `SEMEL` ("semelfactive"), an Aktsionsart
    feature, comes before `ANIM` ("animate").
-   Any language-specific (e.g., `LGSPEC`) tags are then placed at the end, in
    their lexicographic order.

This ordering is enforced by [`um_canonicalize`](um_canonicalize.py); running
this tool on its own output should print "0 feature bundles canonicalized" to
STDERR.

See [`CHANGES.md`](CHANGES.md) for changes to the schema since UniMorph 2.0.

Contributing
------------

The list of features in [`tags.yaml`](tags.yaml) has some known gaps. See
[`CONTRIBUTING.md`](CONTRIBUTING.md) for information about how to submit
improvements to this file.

Author
------

These tools were created by [Kyle Gorman](mailto:kylebgorman@gmail.com).


CHANGES.md
The following featural changes have been made since:

    J. Sylak-Glassman. 2016. The composition and use of the universal morphological feature schema (UniMorph schema). Technical report, Department of Computer Science, Johns Hopkins University.

1.  There are two features denoting "proximate": `PRX` for Person, and `PROX`
    for Case and Deixis. The use of the latter for two categories of feature is
    a problem for automatic canonicalization. Therefore we:

    -   Use `PRX` for Case
    -   Use `PROX` for Deixis
    -   Use `PROXI` for Person

For consistency, we also rename the Person feature `OBV` (obviative) to `OBVI`.

2.  Features with trailing integer do not sort lexicographically if the integer
    is greater than 9, which is the case for both the `BANT*` family of noun
    class features and for the `LGSPEC*` family. Therefore, we pad these two
    families using "%02d" so that they read:

    -   `BANT02` instead of the older `BANT2`
    -   `LGSPEC08` instead of the older `LGSPEC8`

etc.

3.  We add support for multiple feature bundles for "portmanteaux". Each feature
    bundle should be separated using `|` and is processed separately.

**Note that this is not supported yet.**


um_canonicalize/canonicalize.py
"""Canonicalizes feature vectors in UniMorph TSV files.

This applies a number of basic transformations (in particular, sorting) to the
feature vector column of UniMorph TSV files.

For additional context see the original UniMorph schema guidelines:

    www.unimorph.org/doc/Sylak-Glassman_2016_-_UniMorph_Schema_User_Guide.pdf

But note that this supersedes said document as of UniMorph 3.0.
"""

__author__ = "Kyle Gorman"


import argparse
import csv
import logging
import pkg_resources
import re
import sys

from typing import Dict

import yaml

# This maps each feature onto a position in the feature order. By retrieving
# these for each feature we can sort (or detect inconsistencies).
#
# We use the following tag ordering conventions:
#
# -   The first tag is the part of speech tag.
# -   The remaining universal tags are then placed in lexicographic order of
#     the category they represent; e.g., `SEMEL` ("semelfactive"), an
#     Aktionsart feature, comes before `ANIM` ("animate"), an animcy feature.
# -   Any language-specific (e.g., `LGSPEC`) tags are then placed at the end,
#     in their lexicographic order.
#
# Therefore the indices here are ascending except for part of speech (for which
# we reserve 0) and language-specific features (which we handle using a regular
# expression).
#
# TODO(feature): Deal with locative case combinations like `IN+ESS`.
# TODO(feature): Deal with the CN_R_MN switch-reference schemata.

LGSPEC = r"LGSPEC(\d{2,})$"

# Added to the integer in LGSPEC, this gives us its index.
LGSPEC_START = 24


def _load_table(path: str) -> Dict[str, int]:
    with open(path, "r") as source:
        spec = yaml.safe_load(source)
    cat_to_index = {cat: i for (i, cat) in enumerate(spec["ordering"])}
    table: Dict[str, int] = {}
    for (cat, tags) in spec["categories"].items():
        index = cat_to_index[cat]
        for tag in tags:
            table[tag] = index
    return table


def main() -> None:
    logging.basicConfig(level="INFO", format="%(levelname)s: %(message)s")
    parser = argparse.ArgumentParser(
        description="Canonicalizes feature vectors in UniMorph TSV files"
    )
    parser.add_argument("input_path")
    args = parser.parse_args()

    table = _load_table(pkg_resources.resource_filename(__name__, "tags.yaml"))

    # This keeps track of how many bundles were rewritten. If run twice
    # the second time the answer should be 0.
    canonicalized = 0

    with open(args.input_path, "r") as source:
        # TODO: these files contain a bunch of blank lines. Probably should
        # strip them out ahead of time but alternatively we could complexify
        # things here.
        tsv_reader = csv.reader(source, delimiter="\t")
        tsv_writer = csv.writer(sys.stdout, delimiter="\t")
        for (n, (lemma, inflection, features_str)) in enumerate(tsv_reader, 1):
            features = features_str.split(";")
            assert features, "Empty feature vector"
            # Keys are natural number indices; values are the features for
            # that slot.
            slots: Dict[int, str] = {}
            for feature in features:
                try:
                    index = table[feature]
                except KeyError:
                    mtc = re.match(LGSPEC, feature)
                    if mtc:
                        index = LGSPEC_START + int(mtc.group(2))
                    else:
                        logging.error(
                            "Unknown feature: %r (file %s, line %d)",
                            feature,
                            args.input_path,
                            n,
                        )
                        exit(1)
                current = slots.get(index)
                if current:
                    logging.error(
                        "Feature conflict: %r and %r (file %s, line %d)",
                        current,
                        feature,
                        args.input_path,
                        n,
                    )
                    exit(1)
                slots[index] = feature
            canonicalized_features = [
                value for (_, value) in sorted(slots.items())
            ]
            if canonicalized_features != features:
                features_str = ";".join(canonicalized_features)
                canonicalized += 1
            tsv_writer.writerow((lemma, inflection, features_str))

    logging.info("%d feature bundles canonicalized", canonicalized)


if __name__ == "__main__":
    main()


um_canonicalize.py
"""Canonicalizes feature vectors in UniMorph TSV files.

This applies a number of basic transformations (in particular, sorting) to the
feature vector column of UniMorph TSV files.

For additional context see the original UniMorph schema guidelines:

    www.unimorph.org/doc/Sylak-Glassman_2016_-_UniMorph_Schema_User_Guide.pdf

But note that this supersedes said document as of UniMorph 3.0.
"""

__author__ = "Kyle Gorman"


import argparse
import csv
import logging
import pkg_resources
import re
import sys

from typing import Dict

import yaml

# This maps each feature onto a position in the feature order. By retrieving
# these for each feature we can sort (or detect inconsistencies).
#
# We use the following tag ordering conventions:
#
# -   The first tag is the part of speech tag.
# -   The remaining universal tags are then placed in lexicographic order of
#     the category they represent; e.g., `SEMEL` ("semelfactive"), an
#     Aktionsart feature, comes before `ANIM` ("animate"), an animcy feature.
# -   Any language-specific (e.g., `LGSPEC`) tags are then placed at the end,
#     in their lexicographic order.
#
# Therefore the indices here are ascending except for part of speech (for which
# we reserve 0) and language-specific features (which we handle using a regular
# expression).
#
# TODO(feature): Deal with locative case combinations like `IN+ESS`.
# TODO(feature): Deal with the CN_R_MN switch-reference schemata.

LGSPEC = r"LGSPEC(\d{2,})$"

# Added to the integer in LGSPEC, this gives us its index.
LGSPEC_START = 24


def _load_table(path: str) -> Dict[str, int]:
    with open(path, "r") as source:
        spec = yaml.safe_load(source)
    cat_to_index = {cat: i for (i, cat) in enumerate(spec["ordering"])}
    table: Dict[str, int] = {}
    for (cat, tags) in spec["categories"].items():
        index = cat_to_index[cat]
        for tag in tags:
            table[tag] = index
    return table


def main() -> None:
    logging.basicConfig(level="INFO", format="%(levelname)s: %(message)s")
    parser = argparse.ArgumentParser(
        description="Canonicalizes feature vectors in UniMorph TSV files"
    )
    parser.add_argument("input_path")
    args = parser.parse_args()

    table = _load_table(
        pkg_resources.resource_filename("um_canonicalize", "tags.yaml")
    )
    # This keeps track of how many bundles were rewritten. If run twice
    # the second time the answer should be 0.
    canonicalized = 0
    with open(args.input_path, "r") as source:
        # TODO: these files contain a bunch of blank lines. Probably should
        # strip them out ahead of time but alternatively we could complexify
        # things here.
        tsv_reader = csv.reader(source, delimiter="\t")
        tsv_writer = csv.writer(sys.stdout, delimiter="\t")
        for (n, (lemma, inflection, features_str)) in enumerate(tsv_reader, 1):
            features = features_str.split(";")
            assert features, "Empty feature vector"
            # Keys are natural number indices; values are the features for
            # that slot.
            slots: Dict[int, str] = {}
            for feature in features:
                try:
                    index = table[feature]
                except KeyError:
                    mtc = re.match(LGSPEC, feature)
                    if mtc:
                        index = LGSPEC_START + int(mtc.group(2))
                    else:
                        logging.error(
                            "Unknown feature: %r (file %s, line %d)",
                            feature,
                            args.input_path,
                            n,
                        )
                        exit(1)
                current = slots.get(index)
                if current:
                    logging.error(
                        "Feature conflict: %r and %r (file %s, line %d)",
                        current,
                        feature,
                        args.input_path,
                        n,
                    )
                    exit(1)
                slots[index] = feature
            canonicalized_features = [
                value for (_, value) in sorted(slots.items())
            ]
            if canonicalized_features != features:
                features_str = ";".join(canonicalized_features)
                canonicalized += 1
            tsv_writer.writerow((lemma, inflection, features_str))

    logging.info("%d feature bundles canonicalized", canonicalized)
